engine/framework/eventloop/runner.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package eventloop
+
+import (
+	"context"
+	gerrors "errors"
+	"fmt"
+	"time"
+
+	"github.com/pingcap/errors"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"github.com/pingcap/tiflow/engine/pkg/clock"
+	derrors "github.com/pingcap/tiflow/engine/pkg/errors"
+)
+
+const (
+	defaultEventLoopTickInterval = 50 * time.Millisecond
+)
+
+// Runner runs a task.
+type Runner[T task] struct {
+	task T
+
+	alreadyRun atomic.Bool
+	label      string
+
+	clk clock.Clock
+}
+
+// NewRunner returns a new Runner.
+func NewRunner[T task](t T, label string) *Runner[T] {
+	return &Runner[T]{
+		task:  t,
+		label: label,
+		clk:   clock.New(),
+	}
+}
+
+// Run is called by the runtime. Cancelling the
+// ctx cancels the task immediately and forcefully.
+func (r *Runner[R]) Run(ctx context.Context) error {
+	if r.alreadyRun.Swap(true) {
+		panic(fmt.Sprintf("duplicate calls to Run: %s", r.label))
```

Comment:
Better to use log.L().Panic

Reply:
log.L().Panic will not put the zapped fields into the panic info, making it harder to debug